cleanlab/internal/token_classification_utils.py

```diff
@@ -73,64 +76,80 @@ def process_token(token: str, replace: list = [("#", "")]) -> str:
         token: str
             token which potentially contains special characters
 
-        replace: list
+        replace: List[Tuple[str, str]]
             list of tuples `(s1, s2)`, where all occurances of s1 are replaced by s2
 
     Returns
     ---------
-        token: str
+        processed_token: str
             processed token whose special character has been replaced
+
+    Note
+    ----
+        Only applies to characters in the original input token.
     """
-    for old, new in replace:
-        token = token.replace(old, new)
-    return token
+    replace_dict = {re.escape(k): v for (k, v) in replace}
+    pattern = "|".join(replace_dict.keys())
+    compiled_pattern = re.compile(pattern)
+    replacement = lambda match: replace_dict[re.escape(match.group(0))]
+    processed_token = compiled_pattern.sub(replacement, token)
+    return processed_token
 
 
-def mapping(entities: list, maps: list) -> list:
+def mapping(entities: List[int], maps: List[int]) -> List[int]:
     """
     Map a list of entities to its corresponding entities
 
     Parameters
     ----------
-        entities: list
+        entities: List[int]
             a list of given entities
 
-        maps:
+        maps: List[int]
             a list of mapped entities, such that the i'th indexed token should be mapped to `maps[i]`
 
     Returns
     ---------
-        mapped_entities:
+        mapped_entities: List[int]
             a list of mapped entities
 
+    Examples
+    --------
+        >>> unique_identities = [0, 1, 2, 3, 4]  # ["O", "B-PER", "I-PER", "B-LOC", "I-LOC"]
+        >>> maps = [0, 1, 1, 2, 2]  # ["O", "PER", "PER", "LOC", "LOC"]
+        >>> mapping(unique_identities, maps)
+        [0, 1, 1, 2, 2]  # ["O", "PER", "PER", "LOC", "LOC"]
+        >>> mapping([0, 0, 4, 4, 3, 4, 0, 2], maps)
+        [0, 0, 2, 2, 2, 2, 0, 1]  # ["O", "O", "LOC", "LOC", "LOC", "LOC", "O", "PER"]
     """
     f = lambda x: maps[x]
     return list(map(f, entities))
 
 
-def merge_probs(probs: np.ndarray, maps: list) -> np.ndarray:
+def merge_probs(probs: np.ndarray, maps: List[int]) -> np.ndarray:  # type: ignore[type-arg]
     """
-    Merges model-predictive probabilities with desied mapping
+    Merges model-predictive probabilities with desired mapping
 
     Parameters
     ----------
-        probs: np.array
+        probs: np.ndarray
             np.array of shape `(N, K)`, where N is the number of tokens, and K is the number of classes for the model
 
-        maps: list
+        maps: List[int]
             a list of mapped index, such that the probability of the token being in the i'th class is mapped to the
             `maps[i]` index. If `maps[i] == -1`, the i'th column of `probs` is ignored. If `np.any(maps == -1)`, the
             returned probability is re-normalized.
 
     Returns
     ---------
-        probs_merged: np.array
+        probs_merged: np.ndarray
```

Comment:
Could we avoid duplicating types in annotations + docstrings? If we omit the types here, does Sphinx automatically fill them in based on the type annotations? If not, can we configure Sphinx in that way?

Reply:
Removing hard-coded tag. Should be auto-tagged with sphinx-autodoc-typehints in a separate PR.